Add regex search support in lookup table

cailafinn opened the pull request:
**Description of work.**

This PR contains under-the-hood work to implement looking up experiment settings by title as well as angle. 

- Add a `titleMatcher` field to `LookupRow` which contains a regex that will be used to check titles
- Modify `LookupRowFinder` to take the title we want to search for and perform the lookup
- `findLookupRow` now returns a copy of the `LookupRow` rather than a pointer to it. This simplifies the code and ownership significantly for minimal performance impact because `LookupRow` is small and there will only ever be a small number of them. It is wrapped in a `boost::optional` which may be `none` if no matches were found.

Several TODOs have been added which will be addressed in the next steps for #33512

**To test:**

The new functionality is not exposed in the interface yet but here are some suggestions to ensure that the current lookup functionality still works as expected:

- Open the ISIS Reflectometry interface
- Set the instrument to INTER 
- In the Runs table, edit the empty row inside the first group with a run number `13460` and angle `0.5`
- Add another row with run number `13460` and angle `2.3`
- Process both rows and note the values returned for dq/q in the table
- Now, on the Experiment Settings tab, in the table enter a lookup row with angle `0.5` and dq/q `0.01`
- Reproess the Runs tab and dq/q should now be updated to `0.01` for the first row but be unchanged for the second
- Back on Experiment Settings, add another lookup row without an angle but with dq/q of `0.02`
- Reprocess the Runs tab. dq/q should be `0.01` for the first row and `0.02` for the second

Refs #33512

*This does not require release notes* because **it contains under-the-hood changes only**

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Add regex field to lookup row

This is where the string from the view will be stored to match runs
against. The title matcher is the regex present (optionally) in each
row of the experiment tab's lookup table

RE #33512

Commit:
Implement searchByTitle function

Implement the function to search for lookup rows by row title.
Change the lookup functions to return copies rather than pointers. This
simplifies code because we can use stl::copy_if and also reduces the
chances of dangling pointers if misused by callers. Lookup rows should
be small so there is a minimal performance hit.

Re #33512

gemmaguest committed:
Use search by title in findLookupRow

Implements a new way to search for a lookup row by a regex. This is
useful for grouping rows by title to determine their settings

Re #33512

fixup! Implement searchByTitle function

Finish implementing regex matcher

gemmaguest committed:
Implement fallback to match on empty regexes

If a match on explcit regexes is not found, then fall back to empty
regexes if there are any. These are slightly different to wildcard rows
because an angle may be specified, whereas a wildcard row will not have
either an angle or regex.

Re #33512

gemmaguest committed:
Remove use of has_value on boost optional

This is not supported in 1.65 on Ubuntu 18.04

Re #33512

DavidFair reviewed (dismissed):
:+1: LGTM, there's some clean-up we will do in a follow-up PR

gemmaguest committed:
Fix failing tests caused by interim code

The validator needs to wrap the optional regex in another optional to
indicate whether validation succeeded (rather than whether the regex was
supplied)

Re #33512

DavidFair reviewed (dismissed):
`boost::optional<boost::optional<boost::regex>>(boost::optional<boost::regex>(boost::none));` will be replaced with a `parseTitle` method, this is simply a placeholder at the moment
